### PR TITLE
[Merged by Bors] - tortoise: evict persisted state without reading it first

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation

It was an inefficiency in my initial implementation, we already have all required data to evict from the database without touching the disk at all.

related: https://github.com/spacemeshos/go-spacemesh/issues/2852

## Changes
-  unify eviction logic in tortoise/state.go
- persist in multiple batches, to avoid high memory consumption in case of rerun 
- evict from disk based on in-memory data

## Test Plan
existing tests and manual tests on the data from network

## TODO
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: tortoise/state.go [hunk 1]
```diff
@@ -90,12 +82,24 @@ func (s *state) Persist() error {
 				b.Reset()
 			}
 		}
+		if err := batch.Write(); err != nil {
+			return fmt.Errorf("write batch: %w", err)
+		}
+		batch.Reset()
 	}
 
+	if err := batch.Put([]byte(namespaceLast), s.Last.Bytes()); err != nil {
```
Comment (countvonzero) [inline tortoise/state.go:91]: what is the difference doing it here as opposed to earlier?

File: tortoise/state.go [hunk 2]
```diff
@@ -161,37 +165,44 @@ func (s *state) Recover() error {
...
+			}
 		}
-		b.Reset()
-		if err := batch.Delete(it.Key()); err != nil {
-			return fmt.Errorf("delete from batch: %w", err)
+		delete(s.BlockOpinionsByLayer, layerToEvict)
+		if err := batch.Write(); err != nil {
+			return fmt.Errorf("write batch for layer %s to db: %w", layerToEvict, err)
 		}
+		batch.Reset()
```
Comment (countvonzero) [inline tortoise/state.go:203]: what is your guideline for calling `batch.Reset()`? data size?

**Review by countvonzero — APPROVED** [review]

File: tortoise/state.go (hunk 1 above)
Comment (dshulyak) [reply tortoise/state.go:91]: when everything was written in a single batch order didn't matter. but if write fails after writing layers tortoise will think that it has more state than it actually has. on the other hand, if votes are updated but layer is not updated it should be ok

File: tortoise/state.go (hunk 2 above)
Comment (dshulyak) [reply tortoise/state.go:203]: pretty much. it should be more efficient to write everything in a single batch, but writing the whole state after rerun was problematic

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**